### PR TITLE
feat: make post funding required for post create and update, make fun…

### PR DESCRIPTION
**Pull request by CaiWillis**
…ding-type mandatory field

TIS21-6914: Make Post Funding mandatory for all Current Posts

### CODE REVIEW

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/validation/PostValidator.java [hunk 1]
```diff
@@ -306,8 +307,16 @@ private Collection<? extends FieldError> checkFunding(PostDTO postDto) {
...
+        } else if (postFundingDto.getFundingType() == null) {
+          fieldErrors.add(
+              new FieldError(POST_DTO_NAME, "fundings",
+              "Post Funding must have a funding type"));
         }
       }
+    } else if (postDto.getStatus() == Status.CURRENT) { //No Post Funding but CURRENT
+      fieldErrors.add(
+          new FieldError(POST_DTO_NAME, "fundings",
+              "Current Posts Must have at least 1 Post Funding"));
```
Comment (hihilary) [inline tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/validation/PostValidator.java:319]: PostDto.status is not required and is populated from post fundings, so I think backend should not rely on the status field from the client.

Comment (CaiWillis) [reply tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/validation/PostValidator.java:319]: Do you have any advice on a good approach for this? It feels odd to be fetching from the DB at this point but I guess that's the only option given that the field is a backend-only managed one?

Comment (hihilary) [reply tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/validation/PostValidator.java:319]: I think the `validate` method is only used by createPost & updatePost methods in PostResource.
And the client should send all the post fundings attached to the post to the backend. So if you want to check if there's one postFunding, you can just check the dto directly.

File: tcs-service/src/test/java/com/transformuk/hee/tis/tcs/service/api/PostResourceIntTest.java [hunk 2]
```diff
@@ -288,27 +311,13 @@ public void initTest() {
     postGrade = createPostGrade(GRADE_ID, PostGradeType.APPROVED, post);
     postSite = createPostSite(SITE_ID, PostSiteType.PRIMARY, post);
     postSpecialty = createPostSpecialty(specialty, PostSpecialtyType.PRIMARY, post);
+    postFundings = createPostFundings(post);
+    postFundings.stream().forEach(pf -> em.persist(pf));
```
[the remark this answers is not in the document]
Comment (CaiWillis) [reply tcs-service/src/test/java/com/transformuk/hee/tis/tcs/service/api/PostResourceIntTest.java:315]: This turned out the be that cause of most of my issues, thanks!

**Review by hihilary — CHANGES_REQUESTED** [review]
1. Do we need to check if the postFunding Set is an empty set?
2. I think we need more tests in PostResourceIntTest to verify the behaviours if postFunding is null (or empty?).

**Comment by CaiWillis** [comment]
> 1. Do we need to check if the postFunding Set is an empty set?
> 2. I think we need more tests in PostResourceIntTest to verify the behaviours if postFunding is null (or empty?).

1. Good catch, I've replaced the annotation with NotEmpty and added integration tests
2. Done https://github.com/Health-Education-England/TIS-TCS/pull/1047/commits/2f5e68377deb3ec88f4d15904ee86b94a62d5a94

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/validation/PostValidator.java [hunk 3]
```diff
@@ -306,6 +306,10 @@ private Collection<? extends FieldError> checkFunding(PostDTO postDto) {
             && !postFundingDto.getEndDate().isAfter(postFundingDto.getStartDate())) {
           fieldErrors.add(new FieldError(POST_DTO_NAME, "fundings",
               "Post funding end date must not be equal to or before start date"));
+        } else if (postFundingDto.getFundingType() == null) {
+          fieldErrors.add(
+              new FieldError(POST_DTO_NAME, "fundings",
+              "Post Funding must have a funding type"));
```
Comment (hihilary) [inline tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/validation/PostValidator.java:312]: I'm a little wondering if this validation in PostValidator is required as Spring takes it over to validate postFunding NotEmpty in the dto.

Comment (CaiWillis) [reply tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/validation/PostValidator.java:313]: Sorry are you referring to line 300 where we check for empty set? We still need to check each individual post funding for a funding type right?

Comment (hihilary) [reply tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/validation/PostValidator.java:313]: Yes, I did misunderstood the ACs.

**Review by hihilary — APPROVED** [review]
